File: code/packages/csharp/pbkdf2/tests/CodingAdventures.Pbkdf2.Tests/Pbkdf2Tests.cs
using Pbkdf2Algorithm = CodingAdventures.Pbkdf2.Pbkdf2;

namespace CodingAdventures.Pbkdf2.Tests;

public sealed class Pbkdf2Tests
{
    [Fact]
    public void Sha1MatchesRfc6070Vectors()
    {
        Assert.Equal(
            "0c60c80f961f0e71f3a9b524af6012062fe037a6",
            Pbkdf2Algorithm.Pbkdf2HmacSha1Hex("password"u8.ToArray(), "salt"u8.ToArray(), 1, 20));
        Assert.Equal(
            "4b007901b765489abead49d926f721d065a429c1",
            Pbkdf2Algorithm.Pbkdf2HmacSha1Hex("password"u8.ToArray(), "salt"u8.ToArray(), 4096, 20));
        Assert.Equal(
            "56fa6aa75548099dcc37d7f03425e0c3",
            Pbkdf2Algorithm.Pbkdf2HmacSha1Hex("pass\0word"u8.ToArray(), "sa\0lt"u8.ToArray(), 4096, 16));
    }

    [Fact]
    public void Sha256MatchesKnownVectorAndCanExtendPastOneBlock()
    {
        var key = Pbkdf2Algorithm.Pbkdf2HmacSha256("passwd"u8.ToArray(), "salt"u8.ToArray(), 1, 64);
        Assert.Equal(
            "55ac046e56e3089fec1691c22544b605f94185216dde0465e68b9d57c20dacbc49ca9cccf179b645991664b39d77ef317c71b845b1e30bd509112041d3a19783",
            Convert.ToHexString(key).ToLowerInvariant());

        Assert.Equal(
            Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), "salt"u8.ToArray(), 1, 32),
            Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), "salt"u8.ToArray(), 1, 64)[..32]);
    }

    [Fact]
    public void Sha512SupportsCustomKeyLengths()
    {
        var full = Pbkdf2Algorithm.Pbkdf2HmacSha512("secret"u8.ToArray(), "nacl"u8.ToArray(), 1, 64);
        var shortKey = Pbkdf2Algorithm.Pbkdf2HmacSha512("secret"u8.ToArray(), "nacl"u8.ToArray(), 1, 32);

        Assert.Equal(64, full.Length);
        Assert.Equal(shortKey, full[..32]);
        Assert.Equal(128, Pbkdf2Algorithm.Pbkdf2HmacSha512("key"u8.ToArray(), "salt"u8.ToArray(), 1, 128).Length);
    }

    [Fact]
    public void HexHelpersMatchByteOutput()
    {
        Assert.Equal(
            Convert.ToHexString(Pbkdf2Algorithm.Pbkdf2HmacSha1("password"u8.ToArray(), "salt"u8.ToArray(), 1, 20)).ToLowerInvariant(),
            Pbkdf2Algorithm.Pbkdf2HmacSha1Hex("password"u8.ToArray(), "salt"u8.ToArray(), 1, 20));
        Assert.Equal(
            Convert.ToHexString(Pbkdf2Algorithm.Pbkdf2HmacSha256("passwd"u8.ToArray(), "salt"u8.ToArray(), 1, 32)).ToLowerInvariant(),
            Pbkdf2Algorithm.Pbkdf2HmacSha256Hex("passwd"u8.ToArray(), "salt"u8.ToArray(), 1, 32));
        Assert.Equal(
            Convert.ToHexString(Pbkdf2Algorithm.Pbkdf2HmacSha512("secret"u8.ToArray(), "nacl"u8.ToArray(), 1, 64)).ToLowerInvariant(),
            Pbkdf2Algorithm.Pbkdf2HmacSha512Hex("secret"u8.ToArray(), "nacl"u8.ToArray(), 1, 64));
    }

    [Fact]
    public void ValidationRejectsInvalidInputs()
    {
        Assert.Throws<ArgumentNullException>(() => Pbkdf2Algorithm.Pbkdf2HmacSha256(null!, "salt"u8.ToArray(), 1, 32));
        Assert.Throws<ArgumentNullException>(() => Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), null!, 1, 32));
        Assert.Throws<ArgumentException>(() => Pbkdf2Algorithm.Pbkdf2HmacSha256([], "salt"u8.ToArray(), 1, 32));
        Assert.Throws<ArgumentOutOfRangeException>(() => Pbkdf2Algorithm.Pbkdf2HmacSha256("pw"u8.ToArray(), "salt"u8.ToArray(), 0, 32));
        Assert.Throws<ArgumentOutOfRangeException>(() => Pbkdf2Algorithm.Pbkdf2HmacSha256("pw"u8.ToArray(), "salt"u8.ToArray(), 1, 0));
        Assert.Throws<ArgumentOutOfRangeException>(() => Pbkdf2Algorithm.Pbkdf2HmacSha256("pw"u8.ToArray(), "salt"u8.ToArray(), 1, (1 << 20) + 1));
    }

    [Fact]
    public void EmptySaltIsAllowedAndEmptyPasswordCanBeExplicitlyAllowed()
    {
        Assert.Equal(32, Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), [], 1, 32).Length);
        Assert.Equal(32, Pbkdf2Algorithm.Pbkdf2HmacSha256([], "salt"u8.ToArray(), 1, 32, allowEmptyPassword: true).Length);
    }

    [Fact]
    public void SaltPasswordAndIterationsAffectOutput()
    {
        var baseKey = Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), "salt"u8.ToArray(), 1, 32);

        Assert.NotEqual(baseKey, Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), "salt2"u8.ToArray(), 1, 32));
        Assert.NotEqual(baseKey, Pbkdf2Algorithm.Pbkdf2HmacSha256("password2"u8.ToArray(), "salt"u8.ToArray(), 1, 32));
        Assert.NotEqual(baseKey, Pbkdf2Algorithm.Pbkdf2HmacSha256("password"u8.ToArray(), "salt"u8.ToArray(), 2, 32));
    }
}

File: code/packages/fsharp/pbkdf2/tests/CodingAdventures.Pbkdf2.Tests/Pbkdf2Tests.fs
namespace CodingAdventures.Pbkdf2.Tests

open System
open Xunit
open CodingAdventures.Pbkdf2.FSharp

module Pbkdf2Tests =
    [<Fact>]
    let ``sha1 matches RFC 6070 vectors`` () =
        Assert.Equal(
            "0c60c80f961f0e71f3a9b524af6012062fe037a6",
            Pbkdf2.pbkdf2HmacSha1Hex "password"B "salt"B 1 20)
        Assert.Equal(
            "4b007901b765489abead49d926f721d065a429c1",
            Pbkdf2.pbkdf2HmacSha1Hex "password"B "salt"B 4096 20)
        Assert.Equal(
            "56fa6aa75548099dcc37d7f03425e0c3",
            Pbkdf2.pbkdf2HmacSha1Hex "pass\000word"B "sa\000lt"B 4096 16)

    [<Fact>]
    let ``sha256 matches known vector and can extend past one block`` () =
        let key = Pbkdf2.pbkdf2HmacSha256 "passwd"B "salt"B 1 64

        Assert.Equal(
            "55ac046e56e3089fec1691c22544b605f94185216dde0465e68b9d57c20dacbc49ca9cccf179b645991664b39d77ef317c71b845b1e30bd509112041d3a19783",
            Convert.ToHexString(key).ToLowerInvariant())

        Assert.Equal<byte array>(
            Pbkdf2.pbkdf2HmacSha256 "password"B "salt"B 1 32,
            (Pbkdf2.pbkdf2HmacSha256 "password"B "salt"B 1 64).[0..31])

    [<Fact>]
    let ``sha512 supports custom key lengths`` () =
        let full = Pbkdf2.pbkdf2HmacSha512 "secret"B "nacl"B 1 64
        let shortKey = Pbkdf2.pbkdf2HmacSha512 "secret"B "nacl"B 1 32

        Assert.Equal(64, full.Length)
        Assert.Equal<byte array>(shortKey, full.[0..31])
        Assert.Equal(128, (Pbkdf2.pbkdf2HmacSha512 "key"B "salt"B 1 128).Length)

    [<Fact>]
    let ``hex helpers match byte output`` () =
        Assert.Equal(
            Convert.ToHexString(Pbkdf2.pbkdf2HmacSha1 "password"B "salt"B 1 20).ToLowerInvariant(),
            Pbkdf2.pbkdf2HmacSha1Hex "password"B "salt"B 1 20)
        Assert.Equal(
            Convert.ToHexString(Pbkdf2.pbkdf2HmacSha256 "passwd"B "salt"B 1 32).ToLowerInvariant(),
            Pbkdf2.pbkdf2HmacSha256Hex "passwd"B "salt"B 1 32)
        Assert.Equal(
            Convert.ToHexString(Pbkdf2.pbkdf2HmacSha512 "secret"B "nacl"B 1 64).ToLowerInvariant(),
            Pbkdf2.pbkdf2HmacSha512Hex "secret"B "nacl"B 1 64)

    [<Fact>]
    let ``validation rejects invalid inputs`` () =
        Assert.Throws<ArgumentNullException>(fun () -> Pbkdf2.pbkdf2HmacSha256 null "salt"B 1 32 |> ignore) |> ignore
        Assert.Throws<ArgumentNullException>(fun () -> Pbkdf2.pbkdf2HmacSha256 "password"B null 1 32 |> ignore) |> ignore
        Assert.Throws<ArgumentException>(fun () -> Pbkdf2.pbkdf2HmacSha256 [||] "salt"B 1 32 |> ignore) |> ignore
        Assert.Throws<ArgumentException>(fun () -> Pbkdf2.pbkdf2HmacSha256 "pw"B "salt"B 0 32 |> ignore) |> ignore
        Assert.Throws<ArgumentException>(fun () -> Pbkdf2.pbkdf2HmacSha256 "pw"B "salt"B 1 0 |> ignore) |> ignore
        Assert.Throws<ArgumentException>(fun () -> Pbkdf2.pbkdf2HmacSha256 "pw"B "salt"B 1 ((1 <<< 20) + 1) |> ignore) |> ignore

    [<Fact>]
    let ``empty salt is allowed and empty password can be explicitly allowed`` () =
        Assert.Equal(32, (Pbkdf2.pbkdf2HmacSha256 "password"B [||] 1 32).Length)
        Assert.Equal(32, (Pbkdf2.pbkdf2HmacSha256AllowEmptyPassword [||] "salt"B 1 32).Length)

    [<Fact>]
    let ``salt password and iterations affect output`` () =
        let baseKey = Pbkdf2.pbkdf2HmacSha256 "password"B "salt"B 1 32

        Assert.NotEqual<byte array>(baseKey, Pbkdf2.pbkdf2HmacSha256 "password"B "salt2"B 1 32)
        Assert.NotEqual<byte array>(baseKey, Pbkdf2.pbkdf2HmacSha256 "password2"B "salt"B 1 32)
        Assert.NotEqual<byte array>(baseKey, Pbkdf2.pbkdf2HmacSha256 "password"B "salt"B 2 32)
